include/eosio/ship_protocol.hpp

```diff
@@ -0,0 +1,761 @@
+
+   struct account_auth_sequence {
+      eosio::name account  = {};
+      uint64_t    sequence = {};
+   };
+
+   EOSIO_REFLECT(account_auth_sequence, account, sequence)
+
+   struct action_receipt_v0 {
+      eosio::name                        receiver        = {};
+      eosio::checksum256                 act_digest      = {};
+      uint64_t                           global_sequence = {};
+      uint64_t                           recv_sequence   = {};
+      std::vector<account_auth_sequence> auth_sequence   = {};
+      eosio::varuint32                   code_sequence   = {};
+      eosio::varuint32                   abi_sequence    = {};
+   };
+
+   EOSIO_REFLECT(action_receipt_v0, receiver, act_digest, global_sequence, recv_sequence, auth_sequence, code_sequence,
+                 abi_sequence)
+
+   using action_receipt = std::variant<action_receipt_v0>;
+
+   struct account_delta {
+      eosio::name account = {};
+      int64_t     delta   = {};
+   };
+
+   EOSIO_REFLECT(account_delta, account, delta)
+
+   struct action_trace_v0 {
+      eosio::varuint32              action_ordinal         = {};
+      eosio::varuint32              creator_action_ordinal = {};
+      std::optional<action_receipt> receipt                = {};
+      eosio::name                   receiver               = {};
+      action                        act                    = {};
+      bool                          context_free           = {};
+      int64_t                       elapsed                = {};
+      std::string                   console                = {};
+      std::vector<account_delta>    account_ram_deltas     = {};
+      std::optional<std::string>    except                 = {};
+      std::optional<uint64_t>       error_code             = {};
+   };
+
+   EOSIO_REFLECT(action_trace_v0, action_ordinal, creator_action_ordinal, receipt, receiver, act, context_free, elapsed,
+                 console, account_ram_deltas, except, error_code)
+
+   struct action_trace_v1 {
+      eosio::varuint32                   action_ordinal         = {};
+      eosio::varuint32                   creator_action_ordinal = {};
+      std::optional<action_receipt>      receipt                = {};
+      eosio::name                        receiver               = {};
+      action                             act                    = {};
+      bool                               context_free           = {};
+      int64_t                            elapsed                = {};
+      std::string                        console                = {};
+      std::vector<account_delta>         account_ram_deltas     = {};
+      std::vector<account_delta>         account_disk_deltas    = {};
+      std::optional<std::string>         except                 = {};
+      std::optional<uint64_t>            error_code             = {};
+      std::optional<eosio::input_stream> return_value           = {};
```

Comment:
The final cut of the action return value in state_history_plugin doesn't use optional. https://github.com/EOSIO/eos/blob/2db44d405629c0e4182e013cfff1d6f0bbe4b130/plugins/state_history_plugin/include/eosio/state_history_plugin/state_history_serialization.hpp#L598

Reply:
changed